src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp

```diff
@@ -0,0 +1,160 @@
+#include "llvm/ADT/ArrayRef.h"
+#include <cstdint>
+#include <src/Dialect/Mlir/IndexExpr.hpp>
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+// This calculates the values that need to be added to the padding in order to
+// simulate the ceil mode
+llvm::SmallVector<int64_t> getCeilConstants(llvm::ArrayRef<int64_t> inputShape,
+    const ONNXMaxPoolSingleOutOpShapeHelper &shapeHelper, int64_t ceilMode) {
+  // This avoids having more if statements when creating the padding const.
+  if (ceilMode == 0)
+    return llvm::SmallVector<int64_t>{0, 0};
+
+  SmallVector<int64_t, 4> kernelShapeVec =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.kernelShape);
+
+  // Get stride and pad vectors.
+  SmallVector<int64_t, 2> stridesVec = shapeHelper.strides;
+  SmallVector<int64_t, 4> padsVec =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.pads);
+
+  // Check if the idiv_check for the output dimentions in
+  // https://www.mlplatform.org/tosa/tosa_spec.html#_max_pool2d has no
+  // remainder. If it has a remainder, we add size(stride) to the end of the
+  // padding dimension to get one dimension up. Height and width need to have
+  // seperate values.
+  int64_t xAxis = 0;
+  if ((inputShape[2] + padsVec[0] + padsVec[2] - kernelShapeVec[0]) %
+      stridesVec[0])
+    xAxis = stridesVec[0];
+
+  int64_t yAxis = 0;
+  if ((inputShape[3] + padsVec[1] + padsVec[3] - kernelShapeVec[1]) %
+      stridesVec[1])
+    yAxis = stridesVec[1];
+
+  return llvm::SmallVector<int64_t>{xAxis, yAxis};
+}
+
+class ONNXMaxPoolSingleOutOpLoweringToTOSA : public ConversionPattern {
+public:
+  ONNXMaxPoolSingleOutOpLoweringToTOSA(MLIRContext *ctx)
+      : ConversionPattern(ONNXMaxPoolSingleOutOp::getOperationName(), 1, ctx) {}
+
+  using OpAdaptor = typename ONNXMaxPoolSingleOutOp::Adaptor;
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    auto maxpoolOp = llvm::cast<ONNXMaxPoolSingleOutOp>(op);
+    Location loc = op->getLoc();
+    OpAdaptor adaptor(operands, op->getAttrDictionary());
+    // Get shape.
+    IndexExprBuilderForTosa createTosaIE(rewriter, loc);
+    ONNXMaxPoolSingleOutOpShapeHelper shapeHelper(op, operands, &createTosaIE);
+    shapeHelper.computeShapeAndAssertOnFailure();
+
+    Value input = adaptor.X();
+
+    // The attributes storage_order and dilations are unsupported
+    mlir::IntegerAttr storageOrder = adaptor.storage_orderAttr();
+    mlir::ArrayAttr dilations = adaptor.dilationsAttr();
+    mlir::ArrayAttr kernelShape = adaptor.kernel_shapeAttr();
+    const int64_t ceilMode = adaptor.ceil_mode();
+
+    if (input.getType().isa<MemRefType>()) {
+      return rewriter.notifyMatchFailure(
+          op, "memrefs as inputs are unsupported by TOSA");
+    }
+    auto inputType = input.getType().cast<TensorType>();
+
+    if (inputType.getShape().size() != 4) {
+      return rewriter.notifyMatchFailure(
```

Comment:
Only for my own education: what happen on failure? Abort, or it revert to not using TOSA and thus use the KRNL implementation?

Reply:
Currently, we fail in our pipeline if ONNX ops are remaining after converting to TOSA. This might be changed in the long run though.